librz/include/rz_arch.h

```diff
@@ -59,6 +61,245 @@ RZ_API bool rz_arch_load_platform_sdb(RZ_NONNULL RzArchPlatformTarget *t, RZ_NON
+ * * Endian: given explicitly in `big_endian`
+ * * Memories: currently always one memory with index 0 bound against IO, with key size given by `mem_key_size` and value size of 8
+ * * Registers: given explicitly in `reg_bindings` or derived from the register profile with `rz_il_reg_binding_derive()`
+ * * Labels: given explicitly in `labels`
+ * * Initial State of Variables: optionally given in `init_state`
+ */
+typedef struct rz_arch_il_config_t {
+	ut32 pc_size; ///< size of the program counter in bits
+	bool big_endian;
+	/**
+	 * Optional null-terminated array of registers to bind to global vars of the same name.
+	 * If not specified, rz_il_reg_binding_derive will be used.
+	 */
+	RZ_NULLABLE const char **reg_bindings;
+	ut32 mem_key_size; ///< address size for memory 0, bound against IO
+	RzPVector /* <RzILEffectLabel> */ labels; ///< global labels, primarily for syscall/hook callbacks
+	RZ_NULLABLE RzArchILInitState *init_state; ///< optional, initial contents for variables/registers, etc.
+	// more information might go in here, for example additional memories, etc.
+} RzArchILConfig;
+
+typedef bool (*RzArchPluginInit)(RZ_NONNULL RzConfig *config, RZ_NONNULL RZ_OUT void **context);
+typedef bool (*RzArchPluginUpdate)(RZ_NONNULL RzConfig *config, RZ_NULLABLE void *context);
+typedef bool (*RzArchPluginFini)(RZ_NULLABLE void *context);
+typedef bool (*RzArchPluginDisassemble)(RzArchOpcode *opcode, const ut8 *buffer, ut32 buffer_size, ut64 offset, void *context);
+typedef bool (*RzArchPluginAssemble)(const char *assembly, RzStrBuf output, ut64 offset, void *context);
+typedef bool (*RzArchPluginMaskOpcode)(const ut8 *buffer, ut32 buffer_size, RzStrBuf output, ut64 offset, void *context);
+typedef ut32 (*RzArchPluginInfo)(RzArchInfo info, RZ_NULLABLE void *context);
+typedef RzList /*<RzSearchKeyword*>*/ *(*RzArchPluginFunctionPreludes)(RZ_NULLABLE void *context);
+
+typedef bool (*RzArchEsil)(void *esil);
+typedef bool (*RzArchEsilLoop)(void *esil, RzArchOpcode *op);
+typedef bool (*RzArchEsilTrap)(void *esil, int trap_type, int trap_code);
+
+typedef RzArchILConfig *(*RzArchSetupIL)(ut32 endianness);
+
+typedef struct rz_arch_plugin_t {
+	const char *name; ///< plugin name.
+	const char *arch; ///< architecture name.
+	const char *cpus; ///< cpus types (comma separated).
+	const char *desc; ///< description name.
+	const char *author; ///< author name.
+	const char *features; ///< architecture features list.
+	const char *platforms; ///< architecture sdb platform name.
+	ut32 bits; ///< architecture supported bits.
+	ut32 endianness; ///< architecture supported endianness.
+	RzArchPluginInit init; ///< initialize plugin.
+	RzArchPluginUpdate update; ///< updates the plugin configuration.
+	RzArchPluginFini fini; ///< deinitialize plugin.
+	RzArchPluginDisassemble disassemble; ///< disassembles an instruction from the provided bytes.
+	RzArchPluginAssemble assemble; ///< assembles an instruction from the provided assembly.
```

Comment:
This needs another callback for plugins which don't disassemble or assemble.
E.g. rz-ghidra or similar.

Because the different plugins need different parameters, we would need a general way to pass arbitrary arguments.
Or we just pass `core`. But in any way, we would need way to pass what ever the plugin needs.